File: NULASpider.PHP/src/Collections/RedisQueue.php
<?php

// namespace nulastudio\Collections;

// use CSRedis\CSRedisClient;
// use nulastudio\Collections\QueueInterface;
// use nulastudio\Threading\LockManager;

// class RedisQueue implements QueueInterface
// {
//     private $lock = '';

//     private $CSRedis;
//     private $key;

//     public function __construct($connString, array $config = [])
//     {
//         $this->lock = strtoupper(md5(uniqid(mt_rand(), true)));

//         // 连接字符串转换
//         $transConfig   = $this->transConfig($connString, $config);
//         $this->CSRedis = new CSRedisClient($transConfig);
//     }

//     private function transConfig($connString, array $config = [])
//     {
//         #warning connString TODO

//         $host   = '';
//         $params = [];

//         // host提取
//         if (isset($config['host'])) {
//             $host = $config['host'];
//             unset($config['host']);
//         }

//         // key提取
//         if (isset($config['key'])) {
//             $this->key = $config['key'];
//             unset($config['key']);
//         }

//         // params拼接
//         foreach ($config as $key => $value) {
//             $params[] = "{$key}={$value}";
//         }

//         return implode(',', array_merge([$host], $params));
//     }

//     public function pop()
//     {
//         $length = $this->count();
//         if (!$length) {
//             throw new QueueException('The Queue is empty.');
//         }
//         LockManager::getLock($this->lock);
//         $val = null;
//         try {
//             $val = $this->CSRedis->LPop($this->key);
//         } catch (\Exception $e) {
//             throw $e;
//         } finally {
//             LockManager::releaseLock($this->lock);
//         }
//         return $val;
//     }
//     public function push($value)
//     {
//         LockManager::getLock($this->lock);
//         try {
//             $this->CSRedis->RPush<string>($this->key, $value);
//         } catch (\Exception $e) {
//             throw $e;
//         } finally {
//             LockManager::releaseLock($this->lock);
//         }
//     }
//     public function exists($value)
//     {
//         #warning TODO
//         return false;
//     }
//     public function peek()
//     {
//         $length = $this->count();
//         if (!$length) {
//             throw new QueueException('The Queue is empty.');
//         }
//         LockManager::getLock($this->lock);
//         $val = null;
//         try {
//             $val = $this->CSRedis->LIndex($length - 1);
//         } catch (\Exception $e) {
//             throw $e;
//         } finally {
//             LockManager::releaseLock($this->lock);
//         }
//         return $val;
//     }
//     public function count()
//     {
//         LockManager::getLock($this->lock);
//         $val = null;
//         try {
//             $val = $this->CSRedis->LLen($this->key);
//         } catch (\Exception $e) {
//             throw $e;
//         } finally {
//             LockManager::releaseLock($this->lock);
//         }
//         return $val;
//     }
//     function empty() {
//         $this->CSRedis->Del($this->key);
//     }
// }

File: NULASpider.CSharp/Collections/RedisQueue.cs
using System.Linq;
using System;
using System.Collections.Generic;
using System.Text;
using CSRedis;
using Pchp.Core;
using Pchp.Library.Spl;
using nulastudio.Util;

namespace nulastudio.Collections
{
    public class RedisQueue : QueueInterface
    {
        private Context ctx;
        private CSRedisClient CSRedis;
        private string key;

        public RedisQueue(Context ctx, string connString, PhpArray exConfig = null)
        {
            this.ctx = ctx;
            var components = nulastudio.Util.RedisHelper.parseConnectionString(connString);
            this.key = components["key"].String;
            this.CSRedis = new CSRedisClient(components["connectionString"].String);
        }

        public PhpValue pop()
        {
            return this.CSRedis.LPop(this.key);
        }
        public bool push(PhpValue value)
        {
            return this.count() == this.CSRedis.RPush(this.key, value.ToString(this.ctx)) - 1;
        }
        public bool exists(PhpValue value)
        {
            // TODO: 性能问题
            var count = this.count();
            if (count == 0)
            {
                return false;
            }
            for (int i = 0; i < Math.Ceiling(count / 500d); i++)
            {
                var start = i * 500;
                var contains = this.CSRedis.LRange(this.key, start, start + 500 - 1).Contains(value.ToString(this.ctx));
                if (contains)
                {
                    return true;
                }
            }
            return false;
        }
        public PhpValue peek()
        {
            return this.CSRedis.LIndex(this.key, 0);
        }
        public long count()
        {
            return this.CSRedis.LLen(this.key);
        }
        public void empty()
        {
            this.CSRedis.Del(this.key);
        }
        public PhpString serialize(PhpValue value)
        {
            return Pchp.Library.PhpSerialization.serialize(this.ctx, default(RuntimeTypeHandle), value);
        }
        public PhpValue unserialize(PhpString str)
        {
            return Pchp.Library.PhpSerialization.unserialize(this.ctx, default(RuntimeTypeHandle), str);
        }
    }
}
